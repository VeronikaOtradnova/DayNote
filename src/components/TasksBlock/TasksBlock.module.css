.wrapper {
  margin-bottom: 20px;
}

@media (min-width: 1000px) {
  .wrapper {
    width: 40%;
    padding-right: 15px;
    border-right: 1px solid var(--light-gray);
  }
}