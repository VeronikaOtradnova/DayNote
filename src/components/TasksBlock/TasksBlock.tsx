import styles from './TasksBlock.module.css';
import { NewTaskForm } from './forms/NewTaskForm/NewTaskForm';
import { TaskItem } from './TaskItem/TaskItem';
import { useTypedSelector } from '../../hooks/useTypedSelector';
import { useState } from 'react';
import { useCloseOnClickOutside } from '../../hooks/useCloseOnClickOutside';
import { PlusBtn } from '../generic/buttons/PlusBtn/PlusBtn';
import { RemoveTasksBtn } from './btns/RemoveTasksBtn/RemoveTasksBtn';
import { EditTaskBtn } from './btns/EditTaskBtn/EditTaskBtn';
import { XBtn } from '../generic/buttons/XBtn/XBtn';
import { XSquareBtn } from '../generic/buttons/XSquareBtn/XSquareBtn';

export function TasksBlock() {
  const { tasks } = useTypedSelector(store => store.task);
  const { currentDay } = useTypedSelector(store => store.day);
  const tasksList = tasks.filter(task => task.day === currentDay);
  const checkedItems = tasksList.filter(task => task.selected)
  const [isForm, setIsForm] = useState(false);

  const formRef = useCloseOnClickOutside(() => setIsForm(false));

  return (
    <div className={`block-wrapper ${styles.wrapper}`}>
      <div className='block-title-wrapper'>
        <h2 className='block-title'>Задачи</h2>

        {
          checkedItems.length === 1 &&
          <EditTaskBtn />
        }

        {checkedItems.length > 0 && <RemoveTasksBtn />}

        {
          isForm ?
            <XSquareBtn 
              onClick={() => setIsForm(false)}
              testId='tasks-block__hide-form-btn'
              bgSize={30}
            />
            :
            <PlusBtn
              onClick={() => setIsForm(!isForm)}
              testId='tasks-block__show-form-btn'
              bgSize={30}
            />
        }
      </div>

      {
        isForm &&
        <div ref={formRef}>
          <NewTaskForm
            hideForm={() => setIsForm(false)}
          />
        </div>
      }

      {
        tasksList && tasksList.length > 0 &&

        <ul className='list'>
          {
            tasks.filter(task => task.day === currentDay).sort((a, b) => a.created - b.created).map(task => <TaskItem key={task.id} task={task} />)
          }
        </ul>
      }
    </div>
  )
}