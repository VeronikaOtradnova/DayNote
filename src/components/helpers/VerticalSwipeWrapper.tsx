import { useState } from "react"

interface IProps {
  children: React.ReactNode,
  onTop?: () => void,
  onBottom?: () => void,
}

export function VerticalSwipeWrapper({children, onTop = () => {}, onBottom = () => {}}: IProps) {
  const [touchStart, setTouchStart] = useState<number | null>(null)
  const [touchEnd, setTouchEnd] = useState<number | null>(null)

  // the required distance between touchStart and touchEnd to be detected as a swipe
  const minSwipeDistance = 50

  const onTouchStart = (e: React.TouchEvent<HTMLDivElement>) => {
    setTouchEnd(null)
    setTouchStart(e.targetTouches[0].clientY)
  }

  const onTouchMove = (e: React.TouchEvent<HTMLDivElement>) => {
    setTouchEnd(e.targetTouches[0].clientY);
  };

  const onTouchEnd = () => {
    if (!touchStart || !touchEnd) return
    const distance = touchStart - touchEnd
    const isTopSwipe = distance > minSwipeDistance
    const isBottomSwipe = distance < -minSwipeDistance
    if (isTopSwipe || isBottomSwipe) {
      isTopSwipe ? onTop() : onBottom()
    }
  }

  return (
    <div 
      onTouchStart={onTouchStart} 
      onTouchMove={onTouchMove} 
      onTouchEnd={onTouchEnd} 
    >
      {children}
    </div>
  )
}