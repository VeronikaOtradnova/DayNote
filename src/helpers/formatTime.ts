export function formatTime(timeInMillis: number) {
  const date = new Date(timeInMillis);
  
  // Получаем часы и минуты из объекта Date
  let hours = date.getUTCHours().toString();
  let minutes = date.getUTCMinutes().toString();
  
  // Форматируем часы и минуты в виде двухзначных чисел
  hours = hours.padStart(2, '0');
  minutes = minutes.padStart(2, '0');
  
  // Возвращаем строку в формате 'hh:mm'
  return `${hours}:${minutes}`;
}