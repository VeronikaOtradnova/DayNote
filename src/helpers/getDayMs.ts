export const getDayMs = () => {
  return 24*60*60*1000;
}