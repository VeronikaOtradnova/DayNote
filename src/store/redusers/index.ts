import { combineReducers } from "redux";
import { dayReducer } from "./dayReducer";
import { calendarReducer } from "./calendarReducer";

export const rootReducer = combineReducers({
  day: dayReducer,
  calendar: calendarReducer
})

export type TRootState = ReturnType<typeof rootReducer>