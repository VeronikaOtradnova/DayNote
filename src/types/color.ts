export enum colors {
  none = 'none',
  green = 'green',
  pink = 'pink',
  gray = 'gray',
  blue = 'blue',
}

export type TColor = colors.green | colors.pink | colors.gray | colors.none | colors.blue;