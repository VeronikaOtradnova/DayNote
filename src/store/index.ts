import { applyMiddleware, createStore } from "redux";
import { rootReducer } from "./redusers";
import { thunk } from "redux-thunk";

const stateFromStorage = localStorage.getItem('reduxState')
export const persistedState = stateFromStorage ? JSON.parse(stateFromStorage): {}

export const store = createStore(rootReducer, persistedState, applyMiddleware(thunk));

export type AppDispatch = typeof store.dispatch;

store.subscribe(()=>{
  localStorage.setItem('reduxState', JSON.stringify(store.getState()))
})