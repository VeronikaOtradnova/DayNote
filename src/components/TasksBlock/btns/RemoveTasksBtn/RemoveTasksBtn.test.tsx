import { act, screen } from "@testing-library/react"
import { TRootState } from "../../../../store/redusers"
import { renderWithRedux, testInitialState } from "../../../../tests/helpers/renderWithRedux"
import { colors } from "../../../../types/color"
import { TasksBlock } from "../../TasksBlock"
import userEvent from "@testing-library/user-event"

const zeroSelectedTaskInitialState: TRootState = {
  ...testInitialState,
  day: {
    currentDay: +(new Date(2020, 0, 1)),
    days: [{
      date: +(new Date(2020, 0, 1)),
      color: colors.gray,
    }]
  },
  task: {
    tasks: [
      {
        created: +(new Date(2020, 0, 1)),
        done: false,
        id: 'testid0',
        text: 'test task 0',
        day: +(new Date(2020, 0, 1)),
        selected: false,
      },
      {
        created: +(new Date(2020, 0, 1)),
        done: false,
        id: 'testid1',
        text: 'test task 1',
        day: +(new Date(2020, 0, 1)),
        selected: false,
      }
    ]
  }
}

const oneSelectedTaskInitialState: TRootState = {
  ...testInitialState,
  day: {
    currentDay: +(new Date(2020, 0, 1)),
    days: [{
      date: +(new Date(2020, 0, 1)),
      color: colors.gray,
    }]
  },
  task: {
    tasks: [
      {
        created: +(new Date(2020, 0, 1)),
        done: false,
        id: 'testid0',
        text: 'test task 0',
        day: +(new Date(2020, 0, 1)),
        selected: true,
      },
      {
        created: +(new Date(2020, 0, 1)),
        done: false,
        id: 'testid1',
        text: 'test task 1',
        day: +(new Date(2020, 0, 1)),
        selected: false,
      }
    ]
  }
}

const twoSelectedTaskInitialState: TRootState = {
  ...testInitialState,
  day: {
    currentDay: +(new Date(2020, 0, 1)),
    days: [{
      date: +(new Date(2020, 0, 1)),
      color: colors.gray,
    }]
  },
  task: {
    tasks: [
      {
        created: +(new Date(2020, 0, 1)),
        done: false,
        id: 'testid0',
        text: 'test task 0',
        day: +(new Date(2020, 0, 1)),
        selected: true,
      },
      {
        created: +(new Date(2020, 0, 1)),
        done: false,
        id: 'testid1',
        text: 'test task 1',
        day: +(new Date(2020, 0, 1)),
        selected: true,
      }
    ]
  }
}

describe('REMOVE-TASKS-BTN TESTING', () => {
  test('Should render btn when more than zero tasks are selected', () => {
    renderWithRedux(<TasksBlock />, {initialState: oneSelectedTaskInitialState});

    const btn = screen.getByTestId('remove-tasks-btn');
    expect(btn).toBeInTheDocument();
    expect(btn).toMatchSnapshot();
  })

  test('Should not render btn when there are no selected tasks', () => {
    renderWithRedux(<TasksBlock />, {initialState: zeroSelectedTaskInitialState});

    expect(screen.queryByTestId('remove-tasks-btn')).not.toBeInTheDocument();
  })

  test('Should render warning-modal after clicking the btn', async () => {
    renderWithRedux(<TasksBlock />, {initialState: twoSelectedTaskInitialState});

    const user = userEvent.setup();
    const btn = screen.getByTestId('remove-tasks-btn');
    
    await act(async () => {
      await user.click(btn);
    })

    const removeModal = screen.getByTestId('remove-tasks-modal');
    expect(removeModal).toBeInTheDocument();
    const removeBtnInWarningModal = screen.getByTestId('remove-tasks-modal__remove-btn');
    expect(removeBtnInWarningModal).toBeInTheDocument();
  })

  test('Should remove tasks after clicking the btn and confirm in modal', async () => {
    const { store } = renderWithRedux(<TasksBlock />, {initialState: twoSelectedTaskInitialState});

    const user = userEvent.setup();
    const btn = screen.getByTestId('remove-tasks-btn');
    
    await act(async () => {
      await user.click(btn);
    })

    const removeBtnInWarningModal = screen.getByTestId('remove-tasks-modal__remove-btn');

    await act(async () => {
      await user.click(removeBtnInWarningModal);
    })

    const state:TRootState = store.getState();
    expect(state.task.tasks.length).toBe(0);
  })
})