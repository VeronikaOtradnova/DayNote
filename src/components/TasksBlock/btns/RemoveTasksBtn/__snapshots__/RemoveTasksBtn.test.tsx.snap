// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REMOVE-TASKS-BTN TESTING Should render btn when more than zero tasks are selected 1`] = `
<button
  class="btn"
  data-testid="remove-tasks-btn"
/>
`;
